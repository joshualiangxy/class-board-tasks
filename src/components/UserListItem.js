import React, { useState } from 'react';

const UserListItem = ({ user, removeUser, addUser, initialPick }) => {
  const { displayName, uid } = user;
  const [pick, setPick] = useState(initialPick);

  const togglePick = () => {
    setPick(!pick);

    if (pick) removeUser(uid);
    else addUser(uid);
  };

  return (
    <div>
      <label>
        <input type="checkbox" checked={pick} onChange={togglePick} />
        {displayName}
      </label>
    </div>
  );
};

export default UserListItem;
