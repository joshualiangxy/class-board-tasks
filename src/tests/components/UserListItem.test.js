import React from 'react';
import { shallow } from 'enzyme';
import UserListItem from '../../components/UserListItem';
import users from '../fixtures/groupUsers';

let wrapper;
const removeUser = jest.fn();
const addUser = jest.fn();

beforeEach(() => {
  jest.clearAllMocks();

  wrapper = shallow(
    <UserListItem
      user={users[0]}
      initialPick={false}
      removeUser={removeUser}
      addUser={addUser}
    />
  );
});

describe('render', () => {
  it('should render UserListItem', () => expect(wrapper).toMatchSnapshot());

  it('should render UserListItem for alt user', () => {
    wrapper = shallow(
      <UserListItem
        user={users[1]}
        initialPick={true}
        removeUser={removeUser}
        addUser={addUser}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});

describe('checkbox', () => {
  it('should toggle the picked status of user', () => {
    wrapper.find('input').prop('onChange')();

    expect(wrapper.find('input').prop('checked')).toBe(true);

    wrapper.find('input').prop('onChange')();

    expect(wrapper.find('input').prop('checked')).toBe(false);
  });

  it('should call addUser if toggling from false to true', () => {
    wrapper.find('input').prop('onChange')();

    expect(addUser).toHaveBeenCalledTimes(1);
    expect(addUser).toHaveBeenLastCalledWith(users[0].uid);
  });

  it('should call removeUser if toggling from true to false', () => {
    wrapper = shallow(
      <UserListItem
        user={users[1]}
        initialPick={true}
        removeUser={removeUser}
        addUser={addUser}
      />
    );

    wrapper.find('input').prop('onChange')();

    expect(removeUser).toHaveBeenCalledTimes(1);
    expect(removeUser).toHaveBeenLastCalledWith(users[1].uid);
  });
});
