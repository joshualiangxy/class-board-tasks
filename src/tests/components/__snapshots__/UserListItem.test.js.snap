// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render UserListItem 1`] = `
<div>
  <input
    checked={false}
    id="testuid"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="testuid"
  >
    Robert
  </label>
</div>
`;

exports[`render should render UserListItem for alt user 1`] = `
<div>
  <input
    checked={true}
    id="testuid2"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    htmlFor="testuid2"
  >
    Peter
  </label>
</div>
`;
